csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,477 @@
+    if (empty_input_axes.empty()) {
+      // Input is not empty, handle like any other op
+      return;
+    }
+    auto avg = wop->outAvg()->as<TensorView>();
+    auto var_sum = wop->outVar()->as<TensorView>();
+    auto N = wop->outN()->as<TensorView>();
+    // The input is empty in some axes. Assert that they are all reduced
+    for (auto ax : empty_input_axes) {
+      auto id = avg->getRootDomain().at(ax);
+      // Input rfactor domain positions correspond to output root positions
+      TORCH_INTERNAL_ASSERT(
+          id->isReduction(),
+          "Found unexpected unreduced empty axis at position ",
+          ax,
+          " in expression ",
+          wop->toString());
+    }
+
+    // Since WelfordOp has multiple outputs, we need to check whether each is
+    // live before replacing it, to avoid replacing a dead output with a live
+    // one. Note that replaceTV will mark the replacement as live automatically.
+    auto shape = noReductionShape(avg);
+    if (isLive(avg)) {
+      auto nan = IrBuilder::create<Double>(
+          std::numeric_limits<double>::quiet_NaN(), avg->getDataType().value());
+      auto nan_tensor = full(shape, nan, avg->getDataType().value());
+      replaceTV(avg, nan_tensor);
+    }
+    if (isLive(var_sum)) {
+      auto new_var_sum = full(
+          shape,
+          fusion()->zeroVal(var_sum->getDataType().value()),
+          var_sum->getDataType().value());
+      replaceTV(var_sum, new_var_sum);
+    }
+    if (isLive(N)) {
+      auto new_N = full(shape, fusion()->zeroVal(), N->getDataType().value());
+      replaceTV(N, new_N);
+    }
+  }
+
+  //! A cat op can have input empty tensors and still output a non-empty
+  //! tensor. This is only possible if there is more than one input, so we
+  //! only need to handle those cases. We find the non-empty inputs to cat
+  //! then replace with another cat (or `set` if n=1).
+  //!
+  //! The `cat` function creates a CatOp object, but its inputs() are not
+  //! the original inputs. Rather, they are the inputs after padding to the
+  //! output extent in the concatenated dimension. Thus, in the IR graph,
+  //! instead of the following:
+  //!
+  //!    T0  T1   T2
+  //!      \  |  /
+  //!       CatOp
+  //!         |
+  //!        T3
+  //!
+  //! a cat is represented as:
+  //!
+  //!    T0    T1    T2
+  //!     |     |     |
+  //!   PadOp PadOp PadOp
+  //!       \   |   /
+  //!         CatOp
+  //!           |
+  //!          T3
+  //!
+  //! If we determine that one of the inputs, T1, is empty in the cat
+  //! dimension, then we rewrite this as:
+  //!
+  //!    T0          T2
+  //!     |           |
+  //!   PadOp       PadOp
+  //!       \       /
+  //!         CatOp
+  //!           |
+  //!          T3
+  //!
+  //! This is done by simply calling the cat() command with only {T0, T2}.
+  void handle(CatOp* cop) final {
+    auto dim = cop->concatenatedDim();
```

Comment:
ditto. if we have empty axis out side of the cat dimension, we should back out from this.

Reply:
Empty uncatted dim should give an empty out which should be dead. Since we mark definitions dead when we mark `Val`s dead (or we should always do this anyway), it should be an error I think to reach this point.